File: leetcode-04/solutions/cpp/0132.cpp
class Solution {
 public:
  int minCut(string s) {
    const int n = s.length();

    vector<int> cut(n);
    vector<vector<bool>> dp(n, vector<bool>(n));

    for (int i = 0; i < n; i++) {
      int min = i;
      for (int j = 0; j <= i; j++)
        if (s[j] == s[i] && (j + 1 > i - 1 || dp[j + 1][i - 1])) {
          dp[j][i] = true;
          min = j == 0 ? 0 : std::min(min, cut[j - 1] + 1);
        }
      cut[i] = min;
    }

    return cut[n - 1];
  }
};

File: leetcode-04/solutions/python3/0132.py
class Solution:
    def minCut(self, s: str) -> int:
        n = len(s)

        cut = [0] * n
        dp = [[False] * n for _ in range(n)]

        for i in range(n):
            min_ = i
            for j in range(i + 1):
                if s[j] == s[i] and (j + 1 > i - 1 or dp[j + 1][i - 1]):
                    dp[j][i] = True
                    min_ = 0 if j == 0 else min(min_, cut[j - 1] + 1)
            cut[i] = min_

        return cut[n - 1]
